### ISSUE TITLE
 Amazon.S3.Util.S3EventNotification.ParseJson Throws : 'Failed to parse json string: Instance of JsonData doesn't hold an long For large File Size Notifications

### ISSUE THREAD

**Issue by umairamin** [aws/aws-sdk-net#1760]
## Description
When we try to call Amazon.S3.Util.S3EventNotification.ParseJson and JSON contain a notification with number in size field we get an exception Amazon.Runtime.AmazonClientException: 'Failed to parse json string: Instance of JsonData doesn't hold an long'

## Reproduction Steps
Call Amazon.S3.Util.S3EventNotification.ParseJson  with following string
{"Records":[{"eventVersion":"2.1","eventSource":"aws:s3","awsRegion":"ca-central-1","eventTime":"2020-12-07T21:04:12.667Z","eventName":"ObjectCreated:Put","userIdentity":{"principalId":"AWS:AAAAAAAAAAAAAAAAAAAAA"},"requestParameters":{"sourceIPAddress":"0.0.0.0"},"responseElements":{"x-amz-request-id":"5E5E5E5E5E5E5E5E","x-amz-id-2":"STRINGREMOVED"},"s3":{"s3SchemaVersion":"1.0","configurationId":"S3AllEvents","bucket":{"name":"bucket-name-removed","ownerIdentity":{"principalId":"DUMMY"},"arn":"arn:aws:s3:::bucket-name-removed"},"object":{"key":"Dummy File.rar","size":2921782177,"eTag":"11111111111111111111111111111111","sequencer":"005FFFFFFFF"}}}]}

### Environment

* SDK Version: 
* Package Version: 
* OS Info: 
* Build Environment 
* Targeted .NET Platform:  

## Resolution
- [ ] :wave: I can/would-like-to implement a fix for this problem myself

---

This is a :bug: bug-report

**Comment by ashishdhingra** [comment]
Upon comparing with the following sample JSON at [Amazon Simple Storage Service: Event message structure](https://docs.aws.amazon.com/AmazonS3/latest/dev/notification-content-structure.html): 
```JSON
{  
   "Records":[  
      {  
         "eventVersion":"2.1",
         "eventSource":"aws:s3",
         "awsRegion":"us-west-2",
         "eventTime":"1970-01-01T00:00:00.000Z",
         "eventName":"ObjectCreated:Put",
         "userIdentity":{  
            "principalId":"AIDAJDPLRKLG7UEXAMPLE"
         },
         "requestParameters":{  
            "sourceIPAddress":"127.0.0.1"
         },
         "responseElements":{  
            "x-amz-request-id":"C3D13FE58DE4C810",
            "x-amz-id-2":"FMyUVURIY8/IgAtTv8xRjskZQpcIZ9KG4V5Wp6S7S/JRWeUWerMUE5JgHvANOjpD"
         },
         "s3":{  
            "s3SchemaVersion":"1.0",
            "configurationId":"testConfigRule",
            "bucket":{  
               "name":"mybucket",
               "ownerIdentity":{  
                  "principalId":"A3NL1KOZZKExample"
               },
               "arn":"arn:aws:s3:::mybucket"
            },
            "object":{  
               "key":"HappyFace.jpg",
               "size":1024,
               "eTag":"d41d8cd98f00b204e9800998ecf8427e",
               "versionId":"096fKKXTRTtl3on89fVO.nfljtsv6qko",
               "sequencer":"0055AED6DCD90281E5"
            }
         }
      }
   ]
}
``` 

and using the below code:
```C#
using Amazon.S3.Util;

namespace S3EventNotificationParseJson_Issue1760
{
    class Program
    {
        static void Main(string[] args)
        {
            string userJsonString = "{\"Records\":[{\"eventVersion\":\"2.1\",\"eventSource\":\"aws:s3\",\"awsRegion\":\"ca-central-1\",\"eventTime\":\"2020-12-07T21:04:12.667Z\",\"eventName\":\"ObjectCreated:Put\",\"userIdentity\":{\"principalId\":\"AWS:AAAAAAAAAAAAAAAAAAAAA\"},\"requestParameters\":{\"sourceIPAddress\":\"0.0.0.0\"},\"responseElements\":{\"x-amz-request-id\":\"5E5E5E5E5E5E5E5E\",\"x-amz-id-2\":\"STRINGREMOVED\"},\"s3\":{\"s3SchemaVersion\":\"1.0\",\"configurationId\":\"S3AllEvents\",\"bucket\":{\"name\":\"bucket-name-removed\",\"ownerIdentity\":{\"principalId\":\"DUMMY\"},\"arn\":\"arn:aws:s3:::bucket-name-removed\"},\"object\":{\"key\":\"Dummy File.rar\",\"size\":2921782177,\"eTag\":\"11111111111111111111111111111111\",\"sequencer\":\"005FFFFFFFF\"}}}]}";

            string docsJsonString = "{\"Records\":[{\"eventVersion\":\"2.1\",\"eventSource\":\"aws:s3\",\"awsRegion\":\"us-west-2\",\"eventTime\":\"1970-01-01T00:00:00.000Z\",\"eventName\":\"ObjectCreated:Put\",\"userIdentity\":{\"principalId\":\"AIDAJDPLRKLG7UEXAMPLE\"},\"requestParameters\":{\"sourceIPAddress\":\"127.0.0.1\"},\"responseElements\":{\"x-amz-request-id\":\"C3D13FE58DE4C810\",\"x-amz-id-2\":\"FMyUVURIY8/IgAtTv8xRjskZQpcIZ9KG4V5Wp6S7S/JRWeUWerMUE5JgHvANOjpD\"},\"s3\":{\"s3SchemaVersion\":\"1.0\",\"configurationId\":\"testConfigRule\",\"bucket\":{\"name\":\"mybucket\",\"ownerIdentity\":{\"principalId\":\"A3NL1KOZZKExample\"},\"arn\":\"arn:aws:s3:::mybucket\"},\"object\":{\"key\":\"HappyFace.jpg\",\"size\":1024,\"eTag\":\"d41d8cd98f00b204e9800998ecf8427e\",\"versionId\":\"096fKKXTRTtl3on89fVO.nfljtsv6qko\",\"sequencer\":\"0055AED6DCD90281E5\"}}}]}";

            S3EventNotification docsS3EventNotification = S3EventNotification.ParseJson(docsJsonString);
            S3EventNotification userS3EventNotification = S3EventNotification.ParseJson(userJsonString);
        }
    }
}
```

the only point of failure I see is the value of the `size`. I changed the `size` value from `2921782177` to `1024` in user's JSON and it worked. Upon debugging looks like the call to method https://github.com/aws/aws-sdk-net/blob/3826c30e6ec9dbf0505eed22e5ae2dbb6dfbcb9f/sdk/src/Services/S3/Custom/Util/S3EventNotification.cs#L366 for `size` key is throwing exception.

For value `1024` it tries to cast it as `int`.
For value `2921782177` (user's case), it tries to cast as `long`. The cast `(long)data[key]` is throwing exception. It appears that the value `2921782177` is `uint` as per `LitJson` library.

The only fix appears to be at this point is to add extra check for `UInt` in `GetValueAsLong()`:
```C#
private static long GetValueAsLong(JsonData data, string key)
{
    if (data[key] != null)
    {
        if (data[key].IsInt)
            return (int)data[key];
        else if (data[key].IsUInt)
            return (uint)data[key];
        else
            return (long)data[key];
    }
    return 0;
}
```

**Comment by ashishdhingra** [comment]
This should be fixed in version `3.5.7.1` of `AWSSDK.S3`
